src/inventory/transaction/action/BeaconPaymentAction.php

```diff
@@ -0,0 +1,122 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\inventory\transaction\action;
+
+use pocketmine\block\Beacon;
+use pocketmine\block\inventory\BeaconInventory;
+use pocketmine\entity\effect\Effect;
+use pocketmine\event\block\BeaconActivateEvent;
+use pocketmine\inventory\transaction\TransactionValidationException;
+use pocketmine\item\VanillaItems;
+use pocketmine\player\Player;
+use function assert;
+use function in_array;
+
+/**
+ * Represents an action involving a beacon consuming input to produce effects.
+ */
+class BeaconPaymentAction extends InventoryAction{
+
+	public function __construct(
+		protected BeaconInventory $inventory,
+		protected Effect $primaryEffect,
+		protected ?Effect $secondaryEffect
+	){
+		parent::__construct(VanillaItems::AIR(), VanillaItems::AIR());
+	}
+
+	/**
+	 * Returns the beacon inventory involved in this action.
+	 */
+	public function getInventory() : BeaconInventory{
+		return $this->inventory;
+	}
+
+	public function validate(Player $source) : void{
+		$input = $this->inventory->getInput();
+		if(!isset(Beacon::ALLOWED_ITEM_IDS[$input->getTypeId()]) || $input->getCount() < 1){
+			throw new TransactionValidationException("Invalid input item");
+		}
+
+		$position = $this->inventory->getHolder();
+		$block = $position->getWorld()->getBlock($position);
+		if (!$block instanceof Beacon) {
+			throw new TransactionValidationException("Target block is not a beacon");
+		}
+
+		$allowedEffects = $block->getAllowedEffect($block->getBeaconLevel());
+		if(!in_array($this->primaryEffect, $allowedEffects, true)){
+			throw new TransactionValidationException("Primary effect provided is not allowed");
+		}
+		if($this->secondaryEffect !== null && !in_array($this->secondaryEffect, $allowedEffects, true)){
```

Comment:
This validation is incomplete. Secondary effect can only be Regeneration or the same as primary effect. Current code allows that a hacked-cleint to send a primary effect as secondary.